### PR TITLE
8303852: current_stack_region() gets called twice unnecessarily

### PR DESCRIPTION
**Pull request by dholmes-ora**
We combine:
```
address os::current_stack_base();
size_t os::current_stack_size();
```
into
```
void os::current_stack_base_and_size(address* stack_base, size_t* stack_size)
```
and so avoid making two underlying system calls. The os/platform specific specializations are handled by the `current_stack_region` calls. It made sense to modify that to export the `base` directly rather than the `bottom`. In doing that change it made sense to standardise on the the style of the code used (variable names etc) to make it easier to see where the 5 different versions were the same and where they differ. Having 5 versions is unfortunate ( 3 BSD: one per arch - zero, x64, aarch64; and 2 linux: zero and everything else), but trying to combine them with ifdefs would be even worse in my opinion.

In theory there should be zero functional changes here.

Testing:
 - Tiers 1-3
 - All our internal builds in tiers 1-5
 - GHA

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303852](https://bugs.openjdk.org/browse/JDK-8303852): current_stack_region() gets called twice unnecessarily (**Enhancement** - P3)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [434fdb99](https://git.openjdk.org/jdk/pull/15321/files/434fdb99473d97e40c384edc7ec17c77c8f1269b)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [4e22dab0](https://git.openjdk.org/jdk/pull/15321/files/4e22dab0e75efb3806510336eb5720280ae6943f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15321/head:pull/15321` \
`$ git checkout pull/15321`

Update a local copy of the PR: \
`$ git checkout pull/15321` \
`$ git pull https://git.openjdk.org/jdk.git pull/15321/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15321`

View PR using the GUI difftool: \
`$ git pr show -t 15321`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15321.diff">https://git.openjdk.org/jdk/pull/15321.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15321#issuecomment-1683390887)

### CODE REVIEW

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -5394,43 +5397,35 @@ static void current_stack_region(address * bottom, size_t * size) {
...
+
     if (os::Linux::adjustStackSizeForGuardPages()) {
       size_t guard_size = 0;
       rslt = pthread_attr_getguardsize(&attr, &guard_size);
       if (rslt != 0) {
         fatal("pthread_attr_getguardsize failed with error = %d", rslt);
       }
-      *bottom += guard_size;
-      *size   -= guard_size;
+      bottom += guard_size;
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp:5412]: So we can check it in the assert at the end.

File: src/hotspot/os/windows/os_windows.cpp [hunk 2]
```diff
@@ -425,40 +425,31 @@ int os::get_native_stack(address* stack, int frames, int toSkip) {
...
-  stack_bottom =  (address)minfo.AllocationBase;
-  stack_size = minfo.RegionSize;
+  stack_bottom = (address)minfo.AllocationBase;
+  size = minfo.RegionSize;
 
   // Add up the sizes of all the regions with the same
   // AllocationBase.
   while (1) {
-    VirtualQuery(stack_bottom+stack_size, &minfo, sizeof(minfo));
+    VirtualQuery(stack_bottom+size, &minfo, sizeof(minfo));
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.cpp]: Done

File: src/hotspot/os/linux/os_linux.cpp [hunk 3]
```diff
@@ -5394,43 +5397,35 @@ static void current_stack_region(address * bottom, size_t * size) {
...
-  address bottom;
-  size_t size;
-  current_stack_region(&bottom, &size);
-  return size;
+void os::current_stack_base_and_size(address* stack_base, size_t* stack_size) {
+  os::Linux::current_stack_region(stack_base, stack_size);
 }
-#endif
+
+
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp]: Done

**Comment by dholmes-ora** [comment]
Thanks for looking at this @tstuefe - as I noted I didn't try to merge things further as I thought it would makes things more complex. But to address your specific points:

> the implementations for zero and non-zero Linux are identical and could be merged.

But as you then state they are not actually identical because zero is missing some of the guard page checks. But zero also has the IA64 code which causes a subtle change in the way size needs to be adjusted. So they are not the same and while they could potentially be merged I much prefer zero code to be kept isolated to zero so that the folks interested in zero can deal with it.

> bsd aarch64 and x64 could be merged - the only difference is a workaround in x64 that is missing on arm, from 8020753

I did wonder about that workaround but it was not something I wanted to touch in this PR. If it is not needed afterall, or actually needed on Aarch64 then that is something for a different RFE (or actual bug if the workaround is needed on aarch64).

> the bsd zero variant is almost equal to the arm variant, so we could merge those.

Again I like to see zero code isolated, but in addition there is nowhere to put such code except in the shared os_bsd.cpp and then have it ifdef'd out for x86. That said, if the x86 and aarch64 should be the same then yes we could have a single shared version. But again future RFE (I will file one).

> And then, since you equalized the interfaces of os::XX::current_stack_region() and os::get_stack_base_and_size, would it be possible to drop the former and implement the latter for the various OSes?

Yes I think it can. The arch specific versions would just be moved to os namespace instead, and no shared implementation just to redirect to the arch-specific ones.

Thanks.

**Comment by tstuefe** [comment]
Still good. Thanks for fixing this.

File: src/hotspot/os/linux/os_linux.cpp [hunk 4]
```diff
@@ -5367,19 +5367,22 @@ bool os::start_debugging(char *buf, int buflen) {
 //    |                        |/
 // P2 +------------------------+ Thread::stack_base()
 //
-// ** P1 (aka bottom) and size (P2 = P1 - size) are the address and stack size
+// ** P1 (aka bottom) and size are the address and stack size
 //    returned from pthread_attr_getstack().
+// ** P2 (aka stack top or base) = P1 - size)
```
Comment (pchilano) [inline src/hotspot/os/linux/os_linux.cpp:5372]: Preexisting, but isn't P2 = P1 + size? Also an extra parenthesis was left at the end.

File: src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp [hunk 5]
```diff
@@ -709,13 +709,15 @@ size_t os::Posix::default_stack_size(os::ThreadType thr_type) {
 //    |                        |/
 // P2 +------------------------+ Thread::stack_base()
 //
-// ** P1 (aka bottom) and size ( P2 = P1 - size) are the address and stack size returned from
-//    pthread_attr_getstack()
+// ** P1 (aka bottom) and size are the address and stack size
+//    returned from pthread_attr_getstack().
+// ** P2 (aka stack top or base) = P1 - size)
```
Comment (pchilano) [inline src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp:714]: Same thing here.

File: src/hotspot/os_cpu/bsd_zero/os_bsd_zero.cpp [hunk 6]
```diff
@@ -176,70 +176,50 @@ size_t os::Posix::default_stack_size(os::ThreadType thr_type) {
...
+  *base = (address) pthread_get_stackaddr_np(self);
+  *size = pthread_get_stacksize_np(self);
+  bottom = *base - *size;
 #elif defined(__OpenBSD__)
   stack_t ss;
   int rslt = pthread_stackseg_np(pthread_self(), &ss);
 
   if (rslt != 0)
-    fatal("pthread_stackseg_np failed with error = " INT32_FORMAT, rslt);
+    fatal("pthread_stackseg_np failed with error = %s", rslt);
```
Comment (pchilano) [inline src/hotspot/os_cpu/bsd_zero/os_bsd_zero.cpp:192]: %d

**Review by pchilano — APPROVED** [review]
Looks good to me.

File: src/hotspot/os_cpu/bsd_zero/os_bsd_zero.cpp (hunk 6 above)
Comment (dholmes-ora) [reply src/hotspot/os_cpu/bsd_zero/os_bsd_zero.cpp]: Fixed - thanks!

File: src/hotspot/os/linux/os_linux.cpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp]: Good catch!

**Comment by dholmes-ora** [comment]
Thanks for the review @pchilano and for catching those typos!

/integrate

**Pull request closed without merging**